subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -383,66 +404,88 @@ int fota_download_start_with_image_type(const char *host, const char *file,
 		return -EINVAL;
 	}
 
-	if (downloading) {
+	if (atomic_test_and_set_bit(&flags, FLAG_DOWNLOADING)) {
 		return -EALREADY;
 	}
 
+	atomic_clear_bit(&flags, FLAG_CLOSED);
+
+	/* Verify if the URI is same as last time, if not, prevent resuming. */
+	if ((!dl_host || strcmp(dl_host, host) != 0) || (!dl_file || strcmp(dl_file, file) != 0)) {
+		atomic_set_bit(&flags, FLAG_NEW_URI);
+	} else {
+		atomic_clear_bit(&flags, FLAG_NEW_URI);
+	}
+	dl_host = host;
+	dl_file = file;
+
 	if (sec_tag != -1 && !is_ip_address(host)) {
 		config.set_tls_hostname = true;
 	}
 
 	socket_retries_left = CONFIG_FOTA_SOCKET_RETRIES;
 
-	strncpy(file_buf, file, sizeof(file_buf) - 1);
-	file_buf[sizeof(file_buf) - 1] = '\0';
-
 #ifdef PM_S1_ADDRESS
 	/* B1 upgrade is supported, check what B1 slot is active,
 	 * (s0 or s1), and update file to point to correct candidate if
 	 * space separated file is given.
 	 */
 	const char *update;
 	bool s0_active;
+	/* I need modifiable copy of the filename so I can split it */
+	static char file_buf[MAX_RESOURCE_LOCATOR_LEN];
+
+	strncpy(file_buf, file, sizeof(file_buf) - 1);
+	file_buf[sizeof(file_buf) - 1] = '\0';
+	dl_file = file_buf;
 
 	err = fota_download_s0_active_get(&s0_active);
 	if (err != 0) {
+		atomic_clear_bit(&flags, FLAG_DOWNLOADING);
 		return err;
 	}
 
 	err = fota_download_parse_dual_resource_locator(file_buf, s0_active, &update);
 	if (err != 0) {
+		atomic_clear_bit(&flags, FLAG_DOWNLOADING);
 		return err;
 	}
 
 	if (update != NULL) {
 		LOG_INF("B1 update, selected file:\n%s", update);
-		file_buf_ptr = update;
+		dl_file = update;
 	}
 #endif /* PM_S1_ADDRESS */
 
 	img_type_expected = expected_type;
 
-	err = download_client_get(&dlc, host, &config, file_buf_ptr, 0);
+	atomic_set_bit(&flags, FLAG_FIRST_FRAGMENT);
+
+	err = download_client_get(&dlc, dl_host, &config, dl_file, 0);
 	if (err != 0) {
+		atomic_clear_bit(&flags, FLAG_DOWNLOADING);
 		download_client_disconnect(&dlc);
 		return err;
 	}
 
-	downloading = true;
-
 	return 0;
 }
 
 int fota_download_init(fota_download_callback_t client_callback)
 {
+	int err;
+	static bool initialized;
+
 	if (client_callback == NULL) {
 		return -EINVAL;
 	}
 
-	int err;
-
 	callback = client_callback;
 
+	if (initialized) {
+		return 0;
+	}
+
```

Comment:
Is this correct? A local variable "initialized" is not set and this is the first usage of it.

Reply:
Yes, it is correct.

[C11 standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1548.pdf) Says in **6.7.9 Initialization** clause 10:
```
 If an object that has static storage duration is not initialized explicitly,
then:
— if it has pointer type, it is initialized to a null pointer;
— if it has arithmetic type, it is initialized to (positive or unsigned) zero;
```
(This same applies to previous versions as well, at least C99 is the same)

Initializing static variables to zero, or false, will generate an [error with checkpatch](https://github.com/zephyrproject-rtos/zephyr/blob/325e35036bac56f6e21e2adcdf37d9fd616a9446/scripts/checkpatch.pl#L4050-L4058).